### PR TITLE
fix: push create workspace UX to templates page

### PR DESCRIPTION
**Pull request by f0ssel**
@kylecarbs and I have discussed wanting to clean up this flow. The main driver is that Users should have more context than just a template name, and we must provide that flow naturally in the product. The flow now becomes:
- Land on workspaces page
- Click "Create from template" and be brought to the templates page
- Select the template
- Click "Create New Workspace" button
- See template already selected for you and no longer a selector

Example:

chrome_Dvgpap4BzV

### CODE REVIEW

**Comment by ketang** [comment]
I *like* the extra context.  The rationale is solid.  I don't like that it adds a step: the "Create New Workspace" button.  This in particular is annoying because the user has already communicated they want to create a workspace.  You're not so much changing the flow as redirecting them to a different starting point.  Can you pass the creation intention state through the templates page and make it so that selecting the template takes them to the last step instead of the template page?

File: site/src/xServices/createWorkspace/createWorkspaceXService.ts [hunk 1]
```diff
@@ -164,21 +117,20 @@ export const createWorkspaceMachine = createMachine(
...
       assignTemplates: assign({
         templates: (_, event) => event.data,
       }),
       assignSelectedTemplate: assign({
-        selectedTemplate: (_, event) => event.template,
+        selectedTemplate: (ctx, event) => {
+          for (const template of event.data) {
+            if (template.name === ctx.templateName) {
+              return template
+            }
```
Comment (Kira-Pilot) [inline site/src/xServices/createWorkspace/createWorkspaceXService.ts:131]: Can we use Array's [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) here?

**Review by Kira-Pilot — APPROVED** [review]
Code looks good!

**Comment by presleyp** [comment]
Should the CreateWorkspacePage stories get updated in light of this?

**Comment by presleyp** [comment]
I don't think this is the ideal UX for this flow. Neither the status quo nor this seems to address what I'm assuming is the crux of the problem - not the need to know every detail about a template, but the need to have some way of choosing between templates. I would suggest a post-launch meeting of frontend people and people with customer insight for really brainstorming how to solve this from first principles, like being open to introducing a totally new design. In the meantime, if it were me, I'd add a link to the view below saying if you don't know how to choose a template, browse the templates page. But I won't stand in the way of merging this PR. 
Screen Shot 2022-06-08 at 10 32 21 AM

**Comment by f0ssel** [comment]
> We could keep the Create New Workspace button and form, but have it open to a library or carousel of template cards that display the most helpful info about a template, like name and description. Clicking the card would highlight and select it, but each could also have a link to expand the card into a full page modal that displays all the template info.

We originally did have some mock ups like this. I think a good example of this is a create workspace page where selecting a template is similar to the GCP Marketplace page in function:
chrome_QQDoIDOsLF

File: site/src/xServices/createWorkspace/createWorkspaceXService.ts (hunk 1 above)
Comment (f0ssel) [reply site/src/xServices/createWorkspace/createWorkspaceXService.ts:131]: yes, for loops are gonna die hard for me lol

**Comment by kylecarbs** [comment]
I think that's a beautiful eventual, but hard to deliver that much context where we are right now. We lack avatars, have descriptions (but no nice way to set them), and don't have a designated author either.

**Comment by f0ssel** [comment]
@presleyp These aren't very pretty but I think I'm totally fine with this concept:

image

image

My biggest issue is I'm not sure what text should be in these places / what I can do to make this look less meh.

**Comment by f0ssel** [comment]
> I think that's a beautiful eventual, but hard to deliver that much context where we are right now. We lack avatars, have descriptions (but no nice way to set them), and don't have a designated author either.

Totally, that would be a long ways away. My goal in this change is to just correct the flow generally with what we have.

**Comment by presleyp** [comment]
I think it's fine for now! @BrunoQuaresma, look okay to you?

**Comment by f0ssel** [comment]
Asher and Dean are also in favor of the current iteration. I'll get this merged if there's no final objections and anyone who wants to revisit is welcome to.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 2]
```diff
@@ -11,10 +10,11 @@ import { CreateWorkspacePageView } from "./CreateWorkspacePageView"
 const CreateWorkspacePage: FC = () => {
   const organizationId = useOrganizationId()
   const [searchParams] = useSearchParams()
-  const preSelectedTemplateName = searchParams.get("template")
+  const templateParam = searchParams.get("template")
+  const templateName = templateParam ? templateParam : ""
   const navigate = useNavigate()
   const [createWorkspaceState, send] = useMachine(createWorkspaceMachine, {
```
Comment (kylecarbs) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:16]: I think this would ideally be a URL param like: "/templates/<name>/workspace"

**Review by kylecarbs — APPROVED** [review]

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 2 above)
Comment (f0ssel) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx]: done in https://github.com/coder/coder/pull/2142/commits/2a191bad73ac902870f3b965ccafdd3eb309f970

**Pull request merged**